src/components/UriViewer/UriViewer.js

```diff
@@ -56,29 +65,106 @@ export const UriViewer = _.flow(
+      } else {
+        setLoadingError(await e.json())
+      }
     }
   }
+
   useOnMount(() => {
     loadMetadata()
   })
 
-  const { size, timeCreated, updated, bucket, name, fileName, accessUrl } = metadata || {}
-  const gsUri = `gs://${bucket}/${name}`
-  const downloadCommand = getDownloadCommand(fileName, gsUri, accessUrl)
+  const TerminalCommand = downloadCommand => h(Fragment, [
+    p({ style: { marginBottom: '0.5rem', fontWeight: 500 } }, [
+      'Terminal download command'
+    ]),
+    pre({
+      style: {
+        display: 'flex',
+        alignItems: 'center',
+        margin: 0,
+        padding: '0 0.5rem',
+        background: colors.light(0.4),
+      }
+    }, [
+      span({
+        style: {
+          overflowX: 'auto',
+          flex: '1 1 0',
+          padding: '1rem 0',
+        },
+        tabIndex: 0
+      }, [downloadCommand || ' ']),
+      h(ClipboardButton, {
+        'aria-label': 'Copy SAS URL to clipboard',
+        disabled: !downloadCommand,
+        style: { marginLeft: '1ch' },
+        text: downloadCommand,
+      })
+    ])
+  ])
+
+  const FailureMessage = loadingError => [loadingError, () => h(Fragment, [
+    div({ style: { paddingBottom: '1rem' } }, [
+      'Error loading data. This file does not exist or you do not have permission to view it.'
+    ]),
+    h(Collapse, { title: 'Details' }, [
+      div({ style: { marginTop: '0.5rem', whiteSpace: 'pre-wrap', fontFamily: 'monospace', overflowWrap: 'break-word' } }, [
+        JSON.stringify(loadingError, null, 2)
+      ])
+    ])
+  ])]
+
+  const LoadingSymbol = uri => h(Fragment, [
+    isGsUri(uri) || isAzureUri(uri) ? 'Loading metadata...' : 'Resolving DRS file...',
+    spinner({ style: { marginLeft: 4 } })
+  ])
+
   if (isAzureUri(uri)) {
+    const { azureSasStorageUrl, fileName, lastModified, size, workspaceId } = azureStorage || {}
+    uri = azureSasStorageUrl || uri
+    const downloadCommand = getDownloadCommand(fileName, uri)
+    const metadata = { fileName, size }
     return h(Modal, {
       onDismiss,
-      title: 'File Details',
+      title: fileName,
       showCancel: false,
       showX: true,
       okButton: 'Done'
-    },
-    [els.cell([
-      els.label('Filename'),
-      els.data(_.last(uri.split('/')).split('.').join('.\u200B')) // allow line break on periods
-    ]), div({ style: { marginTop: '2rem', fontSize: 14 } }, ['Download functionality for Azure files coming soon'])])
+    }, [
+      Utils.cond(
+        FailureMessage(loadingError),
+        [azureStorage, () => h(Fragment, [
+          azureStorage.lastModified && els.cell([
+            els.label('Last Modified'),
+            els.data(new Date(lastModified).toLocaleString())
+          ]),
+          els.cell([els.label('File size'), els.data(filesize(size))]),
+          h(UriDownloadButton, { uri, metadata, accessUrl, workspaceId }),
+          TerminalCommand(downloadCommand),
+        ])],
+        LoadingSymbol(uri)
```

Comment:
Each argument to `Utils.cond` should be either an array of `[boolean, function]` or a function.